src/main/java/org/breedinginsight/services/TraitUploadService.java

```diff
@@ -182,12 +184,47 @@ public ProgramUpload updateTraitUpload(UUID programId, CompletedFileUpload file,
                 }
             }
 
-            isDup = isDup || this.traitValidator.hasAbbreviation(dupAbbrevTraits, trait, true);
             trait.setIsDup(isDup);
         }
 
         return traits;
     }
+
+    private List<Trait> markScaleNames(List<Trait> traits){
+        for(Trait trait: traits){
+            Scale scale = trait.getScale();
+            if(scale != null){
+                scale.setScaleName( calcDefaultScaleName(trait) );
+            }
+        }
+        return traits;
+    }
+
+    private String calcDefaultScaleName(Trait trait){
+        Scale scale = trait.getScale();
+        if(scale != null){
+            String scaleName = scale.getScaleName();
+            // if scaleName already exist, use it
+            if(scaleName != null) {
+                return scaleName;
+            }
+            else {
+                DataType datatype = scale.getDataType();
+                if( datatype != null ) {
+                    // if the dataType is not Numerical, use the dataType value
+                    if(datatype!=DataType.NUMERICAL){
```

Comment:
Does this mean that if the datatype is `NUMERICAL` that there will not be a scale name derived?

Reply:
Correct, the user is requried to provide the _scale name_ (aka _unit_) if the datatype is NUMERIC.